docs/src/main/asciidoc/websockets-next-reference.adoc

```diff
@@ -0,0 +1,546 @@
+* At most one `@OnTextMessage` method: Handles the connected client's text messages.
+* At most one `@OnBinaryMessage` method: Handles the binary messages the connected client sends.
+* At most one `@OnOpen` method: Invoked when a client connects to the WebSocket.
+* At most one `@OnClose` method: Executed upon the client disconnecting from the WebSocket.
+
+Only some endpoints need to include all methods.
+However, it must contain at least `@On[Text|Binary]Message` or `@OnOpen`.
+
+An error is thrown at build time if any endpoint violates these rules.
+The inner classes representing sub-websockets adhere to the same guidelines.
+
+IMPORTANT: Any methods annotated with `@OnTextMessage`, `@OnBinaryMessage`, `@OnOpen`, and `@OnClose` outside a WebSocket endpoint are considered erroneous and will result in the build failing with an appropriate error message.
+
+== Processing messages
+
+Method receiving messages from the client are annotated with `@OnTextMessage` or `@OnBinaryMessage`.
+
+`OnTextMessage` are invoked for every _text_ message received from the client.
+`OnBinaryMessage` are invoked for every _binary_ message the client receives.
+
+=== Invocation Rules
+
+When invoking these annotated methods, the _session_ scope linked to the WebSocket connection remains active.
+In addition, the request scope is active until the completion of the method (or until it produces its result for async and reactive methods).
+
+Quarkus WebSocket Next supports _blocking_ and _non-blocking_ logic, akin to Quarkus REST, determined by the method signature and additional annotations such as `@Blocking` and `@NonBlocking`.
+
+Here are the rules governing execution:
+
+* Non-blocking methods must execute on the connection's event loop.
+* Methods annotated with `@RunOnVirtualThread` are considered blocking and should execute on a virtual thread.
+* Blocking methods must execute on a worker thread if not annotated with `@RunOnVirtualThread`.
+* When `@RunOnVirtualThread` is employed, each invocation spawns a new virtual thread.
+* Methods returning `CompletionStage` and `Uni` are considered non-blocking
+* Methods returning `Multi` are considered non-blocking and must be subscribed to, except if they return their own `Multi`.
+* Methods returning `void` or plain objects are considered blocking.
+
+=== Parameters
+
+These methods can accept parameters in two formats:
+
+* The message object (of any type).
+* A `Multi<X>` with X as the message type.
+* Any other parameters should be flagged as errors.
+
+The message object represents the data sent and can be accessed as either raw content (`String`, `JsonObject`, `JsonArray`, `Buffer` or `byte[]`) or deserialized high-level objects, which is the recommended approach.
+
+When receiving a `Multi`, the method is invoked once per connection, and the provided `Multi` receives the items transmitted by this connection.
+The method must subscribe to the `Multi` to receive these items (or return a Multi).
```

Comment:
TBH I find the `Multi` param use case quite confusing. First, the method is not executed when a message is received but immediately after the connection is open and second you must not forget to subscribe to the `Multi` (unless returning it).

Maybe we should introduce a separate annotation to consume messages using `Multi`? Something like `@TextMessageStream`?

Reply:
I'm not sure we need to introduce a new annotation. Reactive messaging typically reuses the same. Same for the rest client.

Reply:
We can see if the users are confused.